### ISSUE TITLE
converting Step's `this.parallel` semantics to `queue-flow`

### ISSUE THREAD

**Issue by pk11** [dfellis/queue-flow#7]
I have the following flow right now implemented with `step`:

``` javascript
return step((function () {
    models.execute(db, 'query', "select field1, field2 from...", this.parallel());
    models.execute(db, 'query', "select field1, field2 from...", this.parallel());
    models.execute(db, 'query', "select field1, field2 from...", this.parallel());
    return models.execute(db, "select field1, field2 from...", this.parallel());
  }), (function (error, pItems, nextItem, previousItem, comments) {
     ///executing second step
  }))
```

where `models.execute` is a custom method that's not following the node convention (ie it's defined with no args while arguments are detected via the `arguments` array)

so my attempt to convert this to queue-flow looked roughly like this (tried with `map` too but did not make a difference):

``` javascript
q( [ "select field1, field2 from..."], [ "select field1, field2 from..."], [ "select field1, field2 from..."],  "select field1, field2 from..."], parallel(4)).mapAsync(function(query, callback) {
    models.execute(db, query, function (e,d) {cb(d)} );
}).toArray(function(arr)) {
    ///array is undefined :(
}); 
```

my expectation was that `arr` would contain the query results in order (ie four results) but instead it was `undefined`. 
So I guess my question is how could one execute async method calls using the callbacks in parallel and then do something with the result?

Thanks in advance.

**Comment by dfellis** [comment]
The first thing I notice is that you're wrapping your strings in a series of arrays each sent in as a separate argument to the `q` function.

You'll want to do this, instead:

``` js
q([
    "select field1, field2 from...",
    "select field1, field2 from...",
    "select field1, field2 from...",
    "select field1, field2 from..."
], parallel(4))
```

The first argument is the name of the queue or the array of data to enqueue, and the second is the optional constructor function. This code is trying to use the second array as the constructor function.

Will write a second comment for the next part.

**Comment by dfellis** [comment]
Onto the next part. queue-flow is a functional-style API. I chose this because such an API can look identical between sync and async, unlike in any other flow-control library, but your code isn't going to look imperative if you're taking full advantage of the library.

Instead of listing the queries hardwired like you have here, I'm going to use a named queue so its easier to type :)

``` js
q('queries', parallel(4)) // queries array goes there instead, or keep it named and q('queries').push(query) new ones in later
    .map(function(query) {
        return [db, query];
    }) // turn the query string into an array of the db and query string
    .node(models.execute, 'errorQueue') // since models.execute is a Node.js-style async method, you can use node here and have errors handled by the 'errorQueue'
    .toArray(console.log); // Dump the output array to the screen.
```

That version should work.

EDIT: And note how the different imperative steps taken are put into separate simple functions rather than in bigger functions? That's what I mean when I say functional-style versus imperative style.

One neat advantage of this is that inserting debug code is _very_ easy:

``` js
q('queries', parallel(4))
    .each(console.log) // Log the incoming queries
    .map(function(query) {
        return [db, query];
    })
    .each(console.log) // Log the outgoing arguments
    .node(models.execute, 'errorQueue')
    .each(console.log) // Log the query results
    .toArray(console.log);
```

You can insert `.each` steps anywhere in the code and not worry about altering the behavior of things (because each guarantees it doesn't touch the original queue), and if you see things repeated often, you can make it a named function and just reference that elsewhere.

EDIT2: Whoops, forgot my lambda-js method doesn't do closures at all, replacing that bit of the code.

**Comment by pk11** [comment]
Thank you - super useful! I really like the FP aspect of queue-flow (in fact that's one of the reasons why I am so interested in it). 

The issue I am having is that `models.execute` is not following the node standard, since the argument list is flexible 

``` javascript
//legacy stuff
exports.execute = function() {
  var args, cb, fn, obj, _i;
  obj = arguments[0], fn = arguments[1], args = 4 <= arguments.length ? __slice.call(arguments, 2, _i = arguments.length - 1) : (_i = 2, []), cb = arguments[_i++];
  if (!(fn != null)) {
    return cb("fn is not defined... " + (util.inspect(fn)), null);
  }
  obj[fn].apply(obj, args).success(function(item) {
    return cb(null, item);
  }).error(function(error) {
    return cb(error, null);
  });
};

```

that's why I tried to work around this using map and callbacks.

it sounds like you are recommending to change `models.execute` to a node style method (which is fine). 
Just out of curiosity, could you please elaborate on what happens in a situation where map's callback is called?

thanks again for your help!

**Comment by dfellis** [comment]
Back from lunch. Call it with `nodeAsync`, then, since it is otherwise following the Node-style API.

**Comment by dfellis** [comment]
As for `map(function(val, callback) { ...`, that's useful when need to work with a totally non-Node or non-Promise style async method, such as `someMethod(arg, function(result) { ...`.

If there's only one argument, then you can literally just use `.map(someMethod)` (or `.mapAsync(someMethod)` if `someMethod.length != 2`).

If your function needs more than one argument and one callback, then you need to do:

``` js
.map(function(val, callback) {
    someMethod(arg1, arg2, val, callback);
})
```

But, if it actually matches that (all of the fixed arguments are at the beginning of the method signature), then you can use `bind` and drop the anonymous function:

``` js
.map(someMethod.bind(this, arg1, arg2)) // or mapAsync if needed
```

**Comment by pk11** [comment]
very cool - thanks for the clarification regarding `map`. 

as for the original issue: `nodeAsync` did work! (interestingly enough, I had to remove `parallel(4)`, otherwise `toArray` was not triggered )

**Comment by dfellis** [comment]
Hmm... that sounds like a bug in parallel-queue-flow. If you provided queue-flow with a fixed list of enqueued items, it's supposed to automatically call `close` and trigger `toArray` (and `reduce` and other reduce-like methods).

If you changed it over to `q('someName').push(item1, item2, item3, item4);` then you'll have to explicitly call `q('someName').close()` for `toArray` to run and it's not a bug. (This is because `toArray` doesn't know when to return an array until its sure that all items to be processed have been processed.)

**Comment by pk11** [comment]
I tried out a few things (like closing the main queue before calling `toArray` etc.) but it looks like `toArray` does not seem to work when using `parallel-queue-flow` (and the `parallel(..)` constructor). I will try to put together a test case for this.

As for my current solution 

```
.map(function (query) {
    return [db, query];
})
.nodeAsync(models.execute, 'error')
.toArray(function (arr) {
     ///second step after queries are run
});
```

does this mean that the second call to `models.execute` won't happen until the first callback is finished?

**Comment by dfellis** [comment]
Yes. queue-flow will run multiple steps in parallel automatically and safely, but running multiple pieces of data on the same step at the same time is not done because arbitrary functions aren't guaranteed to be commutative.

What happens if you use `sloppy`, instead?

**Comment by pk11** [comment]
I see - thanks for the explanation.

as for `sloppy`

```
q(['select...','select...'], sloppy)
  .map(function (query) {
    return [db,  query];
  })
  .nodeAsync(models.execute, 'error')
  .toArray(function (arr) {
    console.log(arr);
  });
```

with `sloppy` toArray is indeed called but it's empty (removing `sloppy` fixes the data issue)

**Comment by pk11** [comment]
interestingly, the snippet bellow seems to work fine. Wonder what's special about my use case.

``` javascript
var q = require('queue-flow');
var fs = require('fs');
var sloppy = require('sloppy-queue-flow')


q(['.'], sloppy)
    .node(fs.readdir, 'error')
    .flatten()
    .map(function(filename) {
        return ['./' + filename, 'utf8'];
    })
    .nodeAsync(fs.readFile, 'error')
    .toArray(console.log)
```

**Comment by dfellis** [comment]
:S So the 'close' event is propagating faster than the callbacks complete. That makes sense and is not at all what the intended behavior is...

Thanks. Time to open another issue.

**Comment by dfellis** [comment]
Hmm... I'm pretty sure I solved the issue in sloppy-queue-flow, but I can't reproduce the issue in parallel-queue-flow. [The test suite already proves parallel-queue-flow closes static queues properly after all data has passed through it](https://github.com/dfellis/parallel-queue-flow/blob/master/test/test.js#L17), which makes me think the issue is on the other side involving your DB access?

I'd like to note that your DB and query are treated as an object and reference to a promise, respectively, though. It looks like you could completely rip out the `models.execute` portion and replace it with this:

``` js
q('queries', parallel(4))
    .promise(function(query) {
        return db[query]();
    }, 'errorQueue')
    .toArray(console.log);
```

I'm not entirely sure of that, though, because `models.execute` is a pretty convoluted function and I can't inspect the db object to see if that's truly the case (and to do that would require some fancy-pants code to intercept function calls on non-existent properties and then execute them as you'd like them to be executed, which just seems too clever for its own good).

**Comment by pk11** [comment]
that legacy `models.execute` method makes a connection to mysql via Sequelize, I will have to check and see  how Sequelize manages db connections, but given the warning here https://github.com/felixge/node-mysql it looks like those parallel calls might be sequential after all (I just inherited this project). Anyway, I am not sure if it's related but I thought it was worth sharing.

**Comment by dfellis** [comment]
Well, a quick check into sequelize shows [that its not that crazy](https://github.com/sdepold/sequelize/blob/master/lib/query-interface.js#L90) so I'm guessing I misread how `models.execute` functions.

I'm pretty sure I fixed the issue with `sloppy`, though. [Can you comment on it?](https://github.com/dfellis/sloppy-queue-flow/issues/1)

**Comment by dfellis** [comment]
I identified a bug in issue #6 and I think a similar sort of bug could exist in sloppy-queue-flow and parallel-queue-flow as the source of your code not working in parallel but working sequentially.

Will look into this, but I don't have failing code I can run on my side. Will first try to reproduce the error.

**Comment by dfellis** [comment]
@pk11 I know it's been a while, but can you confirm that this fix in parallel-queue-flow works for you?

**Comment by pk11** [comment]
@dfellis I am going to test it with the latest version and get back to you. Cheers, peter

**Comment by pk11** [comment]
Yeah. I can confirm that it's fixed now.

**Comment by dfellis** [comment]
![party](https://f.cloud.github.com/assets/765551/451387/8ae288a0-b2aa-11e2-9e7e-9506f1105db4.gif)

Closing this issue, and the issue in parallel-queue-flow after I publish it.
